File: app/src/main/java/com/tistory/s1s1s1/inu_contract/DBHelper.java
package com.tistory.s1s1s1.inu_contract;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Administrator on 2016-12-23.
 */

public class DBHelper extends SQLiteOpenHelper {

    public DBHelper(Context context, String name, SQLiteDatabase.CursorFactory factory, int version){
        super(context, name, factory, version);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL("CREATE TABLE CONTRACT(_id INTEGER PRIMARY KEY AUTOINCREMENT, part TEXT, dpart TEXT, position TEXT, name TEXT, task TEXT, phone TEXT, email TEXT, favorite INTEGER);");
        // 소속, 상세소속, 직위, 이름, 직무, 전화번호, 이메일
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        db.execSQL("DROP TABLE CONTRACT");
        onCreate(db);
    }

    public void insert(String part, String dpart, String position, String name, String task, String phone, String email, int favorite){
        SQLiteDatabase db = getWritableDatabase();
        db.execSQL("INSERT INTO CONTRACT VALUES(null, '"+part+"', '"+dpart+"', '"+position+"', '"+name+"', '"+task+"', '"+phone+"', '"+email+"', "+favorite+");");
        db.close();
    }

    public void insertArr(ArrayList<Contract> contracts){
        SQLiteDatabase db = getWritableDatabase();
        for(Contract contract:contracts){
            String part = contract.part;
            String dpart = contract.dpart;
            String position = contract.position;
            String name = contract.name;
            String task = contract.task;
            String phone = contract.phone;
            String email = contract.email;
            int favorite = contract.favorite;
            db.execSQL("INSERT INTO CONTRACT VALUES(null, '"+part+"', '"+dpart+"', '"+position+"', '"+name+"', '"+task+"', '"+phone+"', '"+email+"', "+favorite+");");
        }
        db.close();
    }

    public void drop(){
        SQLiteDatabase db = getWritableDatabase();
        db.execSQL("DROP TABLE CONTRACT");
        db.close();
    }

    public void delete(){
        SQLiteDatabase db = getWritableDatabase();
        db.execSQL("DELETE FROM CONTRACT");
        db.close();
    }

    public int getCount(){
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM CONTRACT", null);
        db.close();
        if(cursor==null) return 0;
        else return cursor.getColumnCount();
    }

    public ArrayList<Contract> getResult(){
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM CONTRACT", null);
        ArrayList<Contract> contracts = new ArrayList();
        Contract contract = null;
        while (cursor.moveToNext()){
            contract = new Contract();
            contract.setPart(cursor.getString(1));
            contract.setDpart(cursor.getString(2));
            contract.setPosition(cursor.getString(3));
            contract.setName(cursor.getString(4));
            contract.setTask(cursor.getString(5));
            contract.setPhone(cursor.getString(6));
            contract.setEmail(cursor.getString(7));
            contract.setFavorite(cursor.getInt(8));

            contracts.add(contract);
        }
        return contracts;
    }

    public ArrayList<String> getPart(){
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT DISTINCT PART FROM CONTRACT ORDER BY PART ASC", null);
        ArrayList<String> parts = new ArrayList<>();
        String part = "";
        while(cursor.moveToNext()){
            part = cursor.getString(0);
            parts.add(part);
        }
        db.close();
        return parts;
    }

    public ArrayList<Contract> getContract(String part){
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM CONTRACT WHERE PART='"+part+"'", null);
        ArrayList<Contract> contracts = new ArrayList<Contract>();
        Contract contract;
        while(cursor.moveToNext()){
            contract = new Contract();
            contract.setData(cursor.getInt(0), cursor.getString(1), cursor.getString(2), cursor.getString(3), cursor.getString(4), cursor.getString(5), cursor.getString(6), cursor.getString(7), cursor.getInt(8));
            contracts.add(contract);
        }
        db.close();
        return contracts;
    }

    public ArrayList<Contract> searchP2(String p){
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT DISTINCT PART FROM CONTRACT WHERE PART LIKE '%"+p+"%'", null);
        ArrayList<Contract> parts = new ArrayList<>();
        Contract contract;
        while(cursor.moveToNext()){
            contract = new Contract();
            contract.setPart(cursor.getString(0));
            parts.add(contract);
        }
        db.close();
        return parts;
    }

    public ArrayList<Contract> searchC(String c){
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM CONTRACT WHERE NAME LIKE'%"+c+"%'", null);
        ArrayList<Contract> contracts = new ArrayList<>();
        Contract contract;
        while(cursor.moveToNext()){
            contract = new Contract();
            contract.setData(cursor.getInt(0), cursor.getString(1), cursor.getString(2), cursor.getString(3), cursor.getString(4), cursor.getString(5), cursor.getString(6), cursor.getString(7), cursor.getInt(8));
            contracts.add(contract);
        }
        db.close();
        return contracts;
    }

    public ArrayList<Contract> searchC(String c, String part){
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM CONTRACT WHERE PART='"+part+"' AND NAME LIKE'%"+c+"%'", null);
        ArrayList<Contract> contracts = new ArrayList<>();
        Contract contract;
        while(cursor.moveToNext()){
            contract = new Contract();
            contract.setData(cursor.getInt(0), cursor.getString(1), cursor.getString(2), cursor.getString(3), cursor.getString(4), cursor.getString(5), cursor.getString(6), cursor.getString(7), cursor.getInt(8));
            contracts.add(contract);
        }
        db.close();
        return contracts;
    }
}

File: app/src/main/java/com/tistory/s1s1s1/inu_contact/Util/DBHelper.kt
/*
 * Copyright (c) 2018. Minjae Son
 */

package com.tistory.s1s1s1.inu_contact.Util

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import com.google.gson.JsonObject
import com.tistory.s1s1s1.inu_contact.Model.Contact
import java.util.*

class DBHelper private constructor(val context: Context, val version: Int) : SQLiteOpenHelper(context, "contact.db", null, version) {
    companion object {
        var instance : DBHelper? = null
        fun getInstance(context : Context, version : Int) : DBHelper{
            val ret = instance ?: DBHelper(context, version)
            instance = ret
            return ret
        }
    }

    val writable: SQLiteDatabase = this.writableDatabase
    val readable: SQLiteDatabase = this.readableDatabase
    val TABLE_ID = "CONTACT"

    val dbCount: Int
        get() {
            val cursor = readable!!.rawQuery("SELECT * FROM CONTACT", null)
            val cnt = cursor!!.count
            cursor.close()
            return cnt
        }

    val part: ArrayList<Contact>
        get() {
            val cursor = readable!!.rawQuery("SELECT DISTINCT PART FROM CONTACT", null)
            val parts = ArrayList<Contact>()
            var contact: Contact
            while (cursor.moveToNext()) {
                contact = Contact()
                contact.part = cursor.getString(0)
                parts.add(contact)
            }
            cursor.close()
            return parts
        }

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL("CREATE TABLE $TABLE_ID(_id INTEGER PRIMARY KEY AUTOINCREMENT, part TEXT, dpart TEXT, position TEXT, name TEXT, task TEXT, phone TEXT, email TEXT, favorite INTEGER);")
        // 소속, 상세소속, 직위, 이름, 직무, 전화번호, 이메일
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL("DROP TABLE CONTACT")
        onCreate(db)
    }

    fun insert(obj: JsonObject) {
        var part = ""
        if (!obj.get("PART").isJsonNull) {
            part = obj.get("PART").asString
        }
        var dpart = ""
        if (!obj.get("DPART").isJsonNull) {
            dpart = obj.get("DPART").asString
        }
        var position = ""
        if (!obj.get("POSITION").isJsonNull) {
            position = obj.get("POSITION").asString
        }
        var name = ""
        if (!obj.get("NAME").isJsonNull) {
            name = obj.get("NAME").asString
        }
        var phone = ""
        if (!obj.get("PHONE").isJsonNull) {
            phone = obj.get("PHONE").asString
            phone = phone.replace("-","")
        }
        val task = ""
        val email = ""
        if(part == "") part ="기타"
        writable!!.execSQL("insert into $TABLE_ID values(null, '$part', '$dpart', '$position', '$name', '$task', '$phone', '$email', 0);")
    }

//    fun drop() {
//        writable!!.execSQL("DROP TABLE CONTACT")
//    }

    fun delete() {
        writable!!.execSQL("DELETE FROM CONTACT")
    }

    fun getContact(part: String): ArrayList<Contact> {
        val cursor = readable!!.rawQuery("SELECT * FROM CONTACT WHERE PART='$part'", null)
        val contacts = ArrayList<Contact>()
        var contact: Contact
        while (cursor.moveToNext()) {
            contact = Contact(cursor.getInt(0), cursor.getString(1), cursor.getString(2), cursor.getString(3), cursor.getString(4), cursor.getString(5), cursor.getString(6), cursor.getString(7), cursor.getInt(8))
            contacts.add(contact)
        }
        cursor.close()
        return contacts
    }

    fun searchPart(part: String): ArrayList<Contact> {
        val cursor = readable!!.rawQuery("SELECT DISTINCT PART FROM CONTACT WHERE PART LIKE '%$part%'", null)
        val parts = ArrayList<Contact>()
        var contact: Contact
        while (cursor.moveToNext()) {
            contact = Contact()
            contact.part = cursor.getString(0)
            parts.add(contact)
        }
        cursor.close()
        return parts
    }

    @JvmOverloads
    fun searchContact(person: String, part :String = ""): ArrayList<Contact> {
        //부서 내부에서 검색할때 part 사용.
        val query = if(part == "") "SELECT * FROM CONTACT WHERE NAME LIKE'%$person%'" else "SELECT * FROM CONTACT WHERE PART='$part' AND NAME LIKE'%$person%'"
        val cursor = readable!!.rawQuery(query, null)
        val contacts = ArrayList<Contact>()
        var contact: Contact
        while (cursor.moveToNext()) {
            contact = Contact(cursor.getInt(0), cursor.getString(1), cursor.getString(2), cursor.getString(3), cursor.getString(4), cursor.getString(5), cursor.getString(6), cursor.getString(7), cursor.getInt(8))
            contacts.add(contact)
        }
        cursor.close()
        return contacts
    }

    @JvmOverloads
    fun searchNumber(phone: String, part :String = ""): ArrayList<Contact> {
        val str = phone.replace("-", "")
        val query = if(part == "") "SELECT * FROM CONTACT WHERE phone LIKE'%$str%'" else "SELECT * FROM CONTACT WHERE PART='$part' AND phone LIKE'%$str%'"
        val cursor = readable!!.rawQuery(query, null)
        val contacts = ArrayList<Contact>()
        var contact: Contact
        while (cursor.moveToNext()) {
            contact = Contact(cursor.getInt(0), cursor.getString(1), cursor.getString(2), cursor.getString(3), cursor.getString(4), cursor.getString(5), cursor.getString(6), cursor.getString(7), cursor.getInt(8))
            contacts.add(contact)
        }
        cursor.close()
        return contacts
    }

}
